src/group_theory/solvable.lean

```diff
@@ -0,0 +1,129 @@
+  intros a h,
+  rw group.mem_conjugates_of_set_iff at h,
+  rcases h with ⟨b, ⟨c, hc, e, he, rfl⟩, d, rfl⟩,
+  exact ⟨d * c * d⁻¹, h₁.conj_mem c hc d, d * e * d⁻¹, h₂.conj_mem e he d, by group⟩,
+end
+
+lemma general_commutator_mono {H₁ H₂ K₁ K₂ : subgroup G} (h₁ : H₁ ≤ K₁) (h₂ : H₂ ≤ K₂) :
+  general_commutator H₁ H₂ ≤ general_commutator K₁ K₂ :=
+begin
+  apply closure_mono,
+  rintros x ⟨p, hp, q, hq, rfl⟩,
+  exact ⟨p, h₁ hp, q, h₂ hq, rfl⟩,
+end
+
+lemma general_commutator_eq_normal_closure_self (H₁ : subgroup G) (H₂ : subgroup G) [H₁.normal]
+  [H₂.normal] : general_commutator H₁ H₂ = normal_closure (general_commutator H₁ H₂) :=
+eq.symm normal_closure_eq_self_of_normal
+
+lemma general_commutator_def' (H₁ H₂ : subgroup G) [H₁.normal] [H₂.normal] :
+  general_commutator H₁ H₂ = normal_closure {x | ∃ (p ∈ H₁) (q ∈ H₂), p * q * p⁻¹ * q⁻¹ = x} :=
+by rw [general_commutator_eq_normal_closure_self, general_commutator,
+  normal_closure_closure_eq_normal_closure]
+
+end general_commutator
+
+section derived_series
+
+variables (G)
+
+/-- The derived series of the group `G`, obtained by starting from the subgroup `⊤` and repeatedly
+  taking the commutator of the previous subgroup with itself for `n` times. -/
+def derived_series (n : ℕ) : subgroup G :=
+nat.rec_on n (⊤ : subgroup G) (λ _ H, general_commutator H H)
+
+@[simp] lemma derived_series_zero : derived_series G 0 = ⊤ := rfl
+
+@[simp] lemma derived_series_succ (n : ℕ) :
+  derived_series G (n + 1) = general_commutator (derived_series G n) (derived_series G n) := rfl
+
+lemma derived_series_normal (n : ℕ) : (derived_series G n).normal :=
+begin
+  induction n with n ih,
+  { exact subgroup.top_normal, },
+  { haveI : (derived_series G n).normal := ih,
+    rw derived_series_succ,
+    exact general_commutator_normal (derived_series G n) (derived_series G n), }
+end
+
+lemma commutator_eq_general_commutator_top_top :
+  commutator G = general_commutator (⊤ : subgroup G) (⊤ : subgroup G) :=
+begin
+  rw [commutator, general_commutator_def'],
+  apply le_antisymm; apply normal_closure_mono,
+  { exact λ x ⟨p, q, h⟩, ⟨p, mem_top p, q, mem_top q, h⟩, },
+  { exact λ x ⟨p, _, q, _, h⟩, ⟨p, q, h⟩, }
+end
+
+lemma commutator_def' : commutator G = subgroup.closure {x : G | ∃ p q, p * q * p⁻¹ * q⁻¹ = x} :=
+begin
+  rw [commutator_eq_general_commutator_top_top, general_commutator],
+  apply le_antisymm; apply closure_mono,
+  { exact λ x ⟨p, _, q, _, h⟩, ⟨p, q, h⟩ },
+  { exact λ x ⟨p, q, h⟩, ⟨p, mem_top p, q, mem_top q, h⟩ }
+end
+
+@[simp] lemma derived_series_one : derived_series G 1 = commutator G :=
+eq.symm $ commutator_eq_general_commutator_top_top G
+
+end derived_series
+
+section solvable
+
+variables (G)
+
+/-- A group `G` is solvable if its derived series is eventually trivial. We use this definition
+  because it's the most convenient one to work with. -/
+def is_solvable : Prop := ∃ n : ℕ, derived_series G n = (⊥ : subgroup G)
```

Comment:
It might be useful to make this a `class`.

Reply:
Does that mean it should be made into a structure? Like should it look like this:
```
class is_solvable : Prop :=
(solvable : ∃ n : ℕ, derived_series G n = ⊥)
```

Reply:
Also, the benefit of this would be that the type inference engine could reason about solvability right?

Reply:
@pglutz You can write `@[class] def` but your definition also works.

Reply:
Is there any advantage to one over the other (i.e. to writing `@[class] def` versus the version I wrote above)?

Reply:
Probably the version that you posted is slightly easier to use. You'll be able to do `cases is_solvable.solvable G` without `unfreezingI` and that kind of headaches.

Reply:
Okay, great. I'll use that version then.